### PR TITLE
✨ Add Mermaid Support to Rich Text Field (Plate)

### PR DESCRIPTION
**Pull request by JackDevAU**
#4734 

> [!NOTE]
> TODO: Write a docs page explaining how to use mermaid in TinaCMS

- Adds mermaid support to Plate editor
	- Preview + edit mermaid diagrams inside a Rich Text Field
	- Saves as ```mermaid codeblock (as per the Standard)
	-  Light mode and dark mode to view mermaid diagram previews better
	- Internal - New type added to Plate - `mermaid`
	- Ability to hide `mermaid` via `toolbarOverride` 
	- Added a test for mermaid diagram
	- Shows errors 
- `<TinaMarkdown />` now has a `mermaid` property that passes the mermaid diagram as a `value` that can then be used to render however the end user wants. Example uses [mermaid npm package](https://www.npmjs.com/package/mermaid)
- Fixes Tooltip's not showing correctly.  

new-mermaid
**Figure: Mermaid support inside Rich Text Field**

image
**Figure: Mermaid Error**

### CODE REVIEW

File: packages/tinacms/src/toolkit/fields/plugins/mdx-field-plugin/plate/components/plate-ui/mermaid-element.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+const LightModeComponent = ({ onToggleMode }) => {
+  const [isLightMode, setIsLightMode] = useState(false)
+
+  const handleToggle = (e) => {
+    e.preventDefault()
+    setIsLightMode((prevMode) => !prevMode)
+  }
+
+  useEffect(() => {
+    //? Note: Adding this class on a <pre/> will remove its base bg styling
```
Comment (ncn-ssw) [inline packages/tinacms/src/toolkit/fields/plugins/mdx-field-plugin/plate/components/plate-ui/mermaid-element.tsx:18]: **nitpick**: Stray '?'
```suggestion
    // Note: Adding this class on a <pre/> will remove its base bg styling
```

**Comment by ncn-ssw** [comment]
**issue**: By default, the component is being rendered in dark mode.

This looks very incongruent with the rest of the current administration interface.

image
**Figure: The dark box with difficult to see lines**

**Comment by ncn-ssw** [comment]
**issue**: The graph is not appearing in the live preview, only its text form.

I expected to see the graph in the live preview when editing was complete on the graph itself.

image
**Figure: The appearance of the block**

**Comment by ncn-ssw** [comment]
**issue**: Saving a mermaid diagram fails with the error message "BlockElement: mermaid is not yet supported"

image  
**Figure: The error dialog shown when the page "Save" button was clicked**

**Comment by ncn-ssw** [comment]
**issue**: The initial form of the Mermaid diagram is a "Syntax Error" message - I expected either a graph or some form of guidance.

The example graph I used below is as follows:

```
graph TD;
    id1(Click the ✏️ edit button at top right of mermaid component) --> id2(Modify this graph);
    id2 --> id3(Click edit button again);
```

image  
**Figure: The "Syntax Error" message from a newly inserted Mermaid graph at the arrow**

**Comment by ncn-ssw** [comment]
**issue**: When there is a syntax error, there is no indication _what_ the error is or any method by which to retrieve the error.

I expected that there would be some sort of useful message.

image  
**Figure: Computer says 'no', doesn't elaborate**

**Comment by ncn-ssw** [comment]
**issue**: Dark-mode is illegible.

I expected that the default color of text would become lighter, rather than remain the same as in light mode.

image
**Figure: Syntax error message is barely visible**

File: packages/tinacms/src/toolkit/fields/plugins/mdx-field-plugin/plate/components/plate-ui/mermaid-element.tsx (hunk 1 above)
Comment (JackDevAU) [reply packages/tinacms/src/toolkit/fields/plugins/mdx-field-plugin/plate/components/plate-ui/mermaid-element.tsx:18]: Ah this is due to an extension that I use - https://marketplace.visualstudio.com/items?itemName=aaron-bond.better-comments 
Makes the comment blue when adding a ?

**Comment by JackDevAU** [comment]
Reviewed the feedback @ncn-ssw 🙏🙏

Added and changed the following, 

- Mermaid diagram's now have a default diagram 
- Made light mode the default 
- Buttons now have a nice blue background and show on hover/click (on mobile) 
- Buttons hide when not hovered
- mermaid diagram defaults to code editor when clicked - instead of showing the syntax warning
- Added a "preview" icon instead of just leaving the edit "pencil" when switching between code and mermaid preview 

image
**Figure: Mermaid default diagram**

new-mermaid
**Figure: New edit and light/dark mode buttons**

**Comment by JackDevAU** [comment]
> issue: The graph is not appearing in the live preview, only its text form.
> I expected to see the graph in the live preview when editing was complete on the graph itself.
> @ncn-ssw 

This is intentional - we are only shipping mermaid support and rendering inside the TinaCMS preview panel. It is up to the end user on how they want to render the mermaid diagram. 

As per my discussion with Jeff we both feel this is the best way as there can be multiple different ways to render mermaid. I'll have what we are using here in the example to test mermaid diagrams documented in the tina.io docs (With code examples). This will also include some other links and methods that can be used to render mermaid diagrams. e.g. remark etc.

File: examples/next-2024/content/posts/hello-world.mdx [hunk 2]
```diff
@@ -1,9 +1,40 @@
...
 Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
+
+```mermaid
+flowchart TD
+    id1(this is an example flow diagram)
+    --> id2(modify me to see changes!)
+    id2
+    --> id3(Click the top button to preview the changes)
+    --> id4(Learn about mermaid diagrams @ mermaid.js.org)
+```
```
Comment (JackDevAU) [inline examples/next-2024/content/posts/hello-world.mdx:26]:
```suggestion

```

we only need 1 diagram for testing!

File: packages/@tinacms/mdx/src/next/stringify/acorn.ts [hunk 3]
```diff
@@ -200,6 +200,7 @@ export function stringifyProps(
         if (field.list) {
           throw new Error(`Rich-text list is not supported`)
         } else {
+          console.log('RICH-TEXT StringifyProps', value)
```
Comment (JackDevAU) [inline packages/@tinacms/mdx/src/next/stringify/acorn.ts:203]:
```suggestion

```

**Review by ncn-ssw — DISMISSED** [review]
🦙 Approved - Excellent work

File: packages/tinacms/src/toolkit/fields/plugins/mdx-field-plugin/plate/components/plate-ui/mermaid-element.tsx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  return null
+}
+
+const DEFAULT_MERMAID_CONFIG = `%% This won't render without implementing a rendering engine (e.g. mermaid on npm)
+flowchart TD
+    id1(this is an example flow diagram) 
+    --> id2(modify me to see changes!)
+    id2 
+    --> id3(Click the top button to preview the changes)
+    --> id4(Learn about mermaid diagrams @ mermaid.js.org)`
```
Comment (JackDevAU) [inline packages/tinacms/src/toolkit/fields/plugins/mdx-field-plugin/plate/components/plate-ui/mermaid-element.tsx:56]:
```suggestion
const DEFAULT_MERMAID_CONFIG = `
flowchart TD
    id1(this is an example flow diagram) 
    --> id2(modify me to see changes!)
    id2 
    --> id3(Click the top button to preview the changes)
    --> id4(Learn about mermaid diagrams @ mermaid.js.org)`
```

File: .changeset/cold-buses-bow.md [hunk 5]
```diff
@@ -0,0 +1,8 @@
+---
+"tinacms": patch
+"@tinacms/mdx": patch
+"@tinacms/schema-tools": patch
```
Comment (JackDevAU) [inline .changeset/cold-buses-bow.md:4]:
```suggestion
"tinacms": patch
"@tinacms/cli": patch
"@tinacms/mdx": patch
"@tinacms/schema-tools": patch
```

**Review by ncn-ssw — DISMISSED** [review]
🦙

**Review by ncn-ssw — APPROVED** [review]
🦙

**Pull request merged**